### PR TITLE
Add an optional param to `Store` to keep the fetches alive even if all downstreams are closed.

### PR DESCRIPTION
**Pull request by eyalgu**
Given the way `Multicaster` is designed, this will not create a hot upstream, but rather allow the existing fetch for complete and deliver that value to the first following downstream that registers.

Note, this PR can be simplified if we requiring buffering is enabled when turning on keeplive, but I didn't go down that route because that feels like the impl is affecting the interface in an unnecessary way.

### CODE REVIEW

**Comment by yigit** [comment]
So my concern about buffering is not about how it is implemented, rather keep alive adds an implicit buffer that may or may not be very clear to the developer.
Making keep-alive work as another buffer means you might get a value that was produced before you've subscribed. This was something we particularly tried to avoid so that code flows that have 
` put some data into server -> fetch updated data` works as desired. 
Adding a keep alive would inadvertently break that. 

I'm not fully sure about forcing buffer either to be honest.

Here are the use cases i want to support:

* I have a `Flow<T>` in my view model that is shared, I want it not to reset on rotation.
   -> in this case, I would already be setting positive buffer as i want data available after rotation

* In Paging 3,0, we have a flow operation that upwraps to another flow, something like this:
`data class A( val events : Flow<B>) `
In some place, we fatmap a fow of A:

**Comment by yigit** [comment]
So my concern about buffering is not about how it is implemented, rather keep alive adds an implicit buffer that may or may not be very clear to the developer.
Making keep-alive work as another buffer means you might get a value that was produced before you've subscribed. This was something we particularly tried to avoid so that code flows that have
`put some data into server -> fetch updated data` works as desired*.
Adding a keep alive would inadvertently break that.

I'm not fully sure about forcing buffer either to be honest.

Here are the use cases i want to support:

I have a Flow<T> in my view model that is shared, I want it not to reset on rotation but still suspend until new UI comes around.
-> in this case, I would already be setting positive buffer as i want data available after rotation

In Paging 3,0, we have a flow operation that unwraps to another flow container, something like this:
```
data class PageData( val events : Flow<PageEvent> ...)
```

For UI, we want to be able to cache the latest PageData (buffer 1) and all events of it (infinite buffer) so that a resubscriber will get all events from the latest PagedData.

To be able to multiplex both of these, we need to manually control the keep alive scope. The result code ends up looking something like: (very experimental code)
https://android-review.googlesource.com/c/platform/frameworks/support/+/1186803/5/paging/common/src/main/kotlin/androidx/paging/CachedPageData.kt#75

Basically, we map each `PagedData` to another one which has multicasted flow, and then scan on that flow to close the previous one manually.


Might be good to list all use cases we can think of to decide what to support. Also, it is easier to open afterwards then remove features :).

File: multicast/src/main/kotlin/com/dropbox/flow/multicast/Multicaster.kt [hunk 1]
```diff
@@ -56,6 +56,11 @@ class Multicaster<T>(
      * it will receive values as well.
      */
     private val piggybackingDownstream: Boolean = false,
+    /**
+     * If true, an active upstream will stay alive even if all downstreams are closed. A downstream
+     * coming in later will receive a value from the live upstream.
+     */
+    private val keepUpstreamAlive: Boolean = false,
```
Comment (yigit) [inline multicast/src/main/kotlin/com/dropbox/flow/multicast/Multicaster.kt]: can we give more granular control over this? (can be a followup CL)
at the least, i would like this to be a time instead of boolean so that it can timeout; can be very handy for usages in a ViewModel where you would want it to suspend for lets say a minute or so before cancelling it to avoid consuming resources (e.g. the real source might be more hot or keeping something else expensive)

**Comment by eyalgu** [comment]
Before making changes to this PR. lets decide on buffering vs no buffering. I don't feel one way or the other so I'll leave it up to the two of you

**Comment by yigit** [comment]
how about we go w/ enforcing buffer for now (which also simplifies it) and then re-consider opening it up in cases where there is no buffer ? It is easier to expand scope later on than reducing it.

**Comment by digitalbuddha** [comment]
@yigit lmk when this is ready, you have more context than me for this.

**Comment by yigit** [comment]
eyalgu@ made a :+1: to my recommendation so i was waiting for that implementation

**Comment by eyalgu** [comment]
updated to enforce buffering

File: store/src/main/java/com/dropbox/android/external/store4/StoreBuilder.kt [hunk 2]
```diff
@@ -36,6 +36,8 @@ interface StoreBuilder<Key, Output> {
     fun scope(scope: CoroutineScope): StoreBuilder<Key, Output>
     fun cachePolicy(memoryPolicy: MemoryPolicy?): StoreBuilder<Key, Output>
     fun disableCache(): StoreBuilder<Key, Output>
+    fun fetchBufferSize(fetchBufferSize: Int): StoreBuilder<Key, Output>
+    fun keepFetcherAlive(keepFetcherAlive: Boolean): StoreBuilder<Key, Output>
```
[the remark this answers is not in the document]
Comment (yigit) [reply store/src/main/java/com/dropbox/android/external/store4/StoreBuilder.kt]: also, i think we do not want to expose buffer here do we?
come to think of it, we need to figure out how this can be used by store. imo, this is a weird case as wanted results from previously fetched data seems very unusual for store and it kind of only makes sense if there is no source of truth.

[the remark this answers is not in the document]
Comment (eyalgu) [reply store/src/main/java/com/dropbox/android/external/store4/StoreBuilder.kt]: I don't have a strong opinion on the API, we can expose it only up the multicaster for now until we decide how to expose it via store

**Comment by eyalgu** [comment]
I won't be able to get to this until next week probably. Feel free to take
over if you'd like otherwise I'll update Monday

**Comment by digitalbuddha** [comment]
no rush, open source moves at it's own pace :-)

File: multicast/src/test/kotlin/com/dropbox/flow/multicast/ChannelManagerTest.kt [hunk 3]
```diff
@@ -141,6 +143,234 @@ class ChannelManagerTest {
...
+                    .map { it.value }
+            }
+            upstream.send("c")
+            upstream.close()
+
+            assertThat(s2.await()).isEqualTo(listOf("b", "c"))
+        }
+
+    @Test
+    fun `GIVEN keepUpstreamAlive AND buffer of size 1 WHEN add two non overlapping downstreams AND a 3rd channel overlapping the 2nd channle AND emit during keepalive THEN 3rd channel receives 1 values`() =
```
[the remark this answers is not in the document]
Comment (eyalgu) [reply multicast/src/test/kotlin/com/dropbox/flow/multicast/ChannelManagerTest.kt]: yeah, that's why I suggested using speak of some more dedicated BDD framework so we could do:
```
class MyTests({
  GIVEN("a state") {
    // set up state
    WHEN("something happens") {
      // make something happened
      THEN("expect a result")
        // verify result
    }
    WHEN("something else happens") {
      // make something else happened
      THEN("expect another result")
        // verify another result
    }
  }
}) 
```

**Comment by digitalbuddha** [comment]
```
Task :multicast:ktlintTestSourceSetCheck FAILED
/home/travis/build/dropbox/Store/multicast/src/test/kotlin/com/dropbox/flow/multicast/ChannelManagerTest.kt:374:1: Unexpected indentation (2) (it should be 4) (cannot be auto-corrected)
/home/travis/build/dropbox/Store/multicast/src/test/kotlin/com/dropbox/flow/multicast/ChannelManagerTest.kt:374:1: Trailing space(s)
"plain" report written to /home/travis/build/dropbox/Store/multicast/build/reports/ktlint/ktlintTestSourceSetCheck.txt
FAILURE: Build failed with an exception.
```
